app/bundles/FormBundle/Model/SubmissionModel.php

```diff
@@ -1048,16 +1050,55 @@ protected function createLeadFromSubmit(Form $form, array $leadFieldMatches, $le
             }
         }
 
-        if ($companyEntity) {
-            unset($data['company']);
-        }
-
         // set the mapped fields
         $this->leadModel->setFieldValues($lead, $data, false, true, true);
 
         // last active time
         $lead->setLastActive(new \DateTime());
 
+        // Set stage.
+        if (!empty($data['stagebyname'])) {
+            $stage = $this->em->getRepository(Stage::class)->findByIdOrName($data['stagebyname']);
+
+            if ($stage instanceof Stage) {
+                $lead->setStage($stage);
+
+                // Add a contact stage change log
+                $lead->stageChangeLogEntry(
+                    $stage,
+                    sprintf('%d:%s', $stage->getId(), $stage->getName()),
+                    $this->translator->trans(
+                        'mautic.stage.import.action.name',
+                        ['%name%' => $this->userHelper->getUser()->getUsername()]
+                    )
+                );
+            } else {
+                $this->logger->warning(sprintf(
+                    'Form: Associating stage failed as %s',
+                    $this->translator->trans(
+                        'mautic.lead.import.stage.not.exists',
+                        ['%id%' => $data['stagebyname']]
+                    )
+                ));
+            }
+        }
+
+        // Set owner
+        $ownerEmailOrId = null;
+        if (!empty($data['ownerbyemail'])) {
+            $ownerEmailOrId = $data['ownerbyemail'];
+        } elseif (!empty($data['ownerbyid'])) {
+            $ownerEmailOrId = $data['ownerbyid'];
+        }
+
+        if ($ownerEmailOrId) {
+            $user = $this->em->getRepository(User::class)->findOneByEmailOrId($ownerEmailOrId);
+
+            if ($user instanceof User) {
+                $lead->setOwner($user);
+            }
+        }
```

Comment:
I don't really like the ORs when you can nicely ask with the actual value like
```suggestion
        $userRepo = $this->em->getRepository(User::class);
        $user = null;
        if (!empty($data['ownerbyemail'])) {
            $user = $userRepo->findOneBy(['email' => $data['ownerbyemail']);
        } elseif (!$user && !empty($data['ownerbyid'])) {
            $user = $userRepo->find($data['ownerbyid']);
        }

        if ($user instanceof User) {
            $lead->setOwner($user);
        }
```
It's shorter and you can even delete the new `findOneByEmailOrId` method. Same can be done with the stage. For some reason the user has 2 values. One for ID, another for email. But stage has 1 value for name and ID.

Reply:
This is super clean, I'll update the PR accordingly. Thanks @escopecz.

Reply:
@escopecz, the `\Mautic\StageBundle\Entity\StageRepository::findByIdOrName` method is called by others, so I'll not refactor that here in this PR. I can take up the cleaning in the new PR.